### ISSUE TITLE
Pages / Directories: Restrict by User Area

### ISSUE THREAD

**Issue by HeyJoel** [cofoundry-cms/cofoundry#187]
We need to be able to restrict a content-editable page or directory to a specific user area. An example of this would be having a Members user area under the directory */members/* with content manageable sub-directories and pages.

We'd also have to think about protected assets here too.

**Comment by HeyJoel** [comment]
A workaround for restricting content managed pages/directories is to customize the routing pipline and add a step to validate the user area.

**NB: This overrides undocumented features that may change in future without warning**

**ValidateUserAreaRoutingStep .cs**
```csharp
using System.Threading.Tasks;
using Cofoundry.Web;
using Microsoft.AspNetCore.Mvc;

/// <summary>
/// This is the routing step that validates the request
/// </summary>
public class ValidateUserAreaRoutingStep : IPageActionRoutingStep
{
    public Task ExecuteAsync(Controller controller, PageActionRoutingState state)
    {
        if (IsRestrictedPath(state.InputParameters.Path)
            && !state.IsCofoundryAdminUser
            && state.AmbientUserContext?.UserArea?.UserAreaCode != MemberUserArea.Code)
        {
            state.Result = new StatusCodeResult(403);
        }

        return Task.CompletedTask;
    }

    private bool IsRestrictedPath(string path)
    {
        if (path == null) return false;

        // Use whatever method you want to determine the restricted path
        return path.StartsWith("members");
    }
}
``` 

**CustomPageActionRoutingStepFactory.cs**
```csharp
using System.Collections.Generic;
using Cofoundry.Web;

/// <summary>
/// A factory that creates a collection of ordered IPageActionRoutingStep to
/// be executed in order during the PageController's Page Action. This determines
/// the routing of dynamic page content in the site.
/// </summary>
public class CustomPageActionRoutingStepFactory : IPageActionRoutingStepFactory
{
    private readonly IEnumerable<IPageActionRoutingStep> _routingSteps;

    public CustomPageActionRoutingStepFactory(
        ICheckSiteIsSetupRoutingStep checkSiteIsSetupRoutingStep,
        IInitStateRoutingStep initStateRoutingStep,
        ITryFindPageRoutingInfoRoutingStep tryFindPageRoutingInfoRoutingStep,
        IValidateEntityEditModeRoutingStep validateEntityEditModeRoutingStep,
        IValidateEditPermissionsRoutingStep validateEditPermissionsRoutingStep,
        IValidateDraftVersionRoutingStep validateDraftVersionRoutingStep,
        IValidateSpecificVersionRoutingRoutingStep validateSpecificVersionRoutingRoutingStep,
        IGetNotFoundRouteRoutingStep getNotFoundRouteRoutingStep,
        IGetPageRenderDataRoutingStep getPageRenderDataRoutingStep,
        ISetCachePolicyRoutingStep setCachePolicyRoutingStep,
        IGetFinalResultRoutingStep getFinalResultRoutingStep
        )
    {
        // Here we set the default routing steps, which are run in the order they are 
        // declared. Each step can be overridden using the DI system if you need to debug it
        // in an implemenetation.

        var routingSteps = new List<IPageActionRoutingStep>()
        {
            checkSiteIsSetupRoutingStep,
            initStateRoutingStep,
            // Here we add our new routing step into the pipeline
            new ValidateUserAreaRoutingStep(),
            tryFindPageRoutingInfoRoutingStep,
            // Alternatively you could add your routing step here 
            // if you needed to know if the route existed
            validateEntityEditModeRoutingStep,
            validateEditPermissionsRoutingStep,
            validateDraftVersionRoutingStep,
            validateSpecificVersionRoutingRoutingStep,
            getNotFoundRouteRoutingStep,
            getPageRenderDataRoutingStep,
            setCachePolicyRoutingStep,
            getFinalResultRoutingStep
        };

        _routingSteps = routingSteps;
    }

    public IEnumerable<IPageActionRoutingStep> Create()
    {
        return _routingSteps;
    }
}
```

**DependencyRegistration.cs**
```csharp
public class DependencyRegistration : IDependencyRegistration
{
    public void Register(IContainerRegister container)
    {
        // Here we simply override the default factory with our own
        container.Register<IPageActionRoutingStepFactory, CustomPageActionRoutingStepFactory>(RegistrationOptions.Override());
    }
}
````

**Comment by HeyJoel** [comment]
This has been implemented as "Access Control" for pages and directories. It will be released in v0.10. More information will be provided in the docs and blog post for v0.10.
